storage/innobase/row/row0merge.cc

```diff
@@ -2325,6 +2468,13 @@ row_merge_read_clustered_index(
 			vers_update_trt = true;
 		}
 
+		if (!charset_converter.convert_if_needed(
+			row, row_heap, old_table, table->s,
+			static_cast<ulong>(n_rows + 1))) {
+			err = DB_TRUNCATED_WRONG_VALUE_FOR_FIELD;
+			break;
+		}
```

Comment:
Do we want this conversion to occur inside InnoDB at all?

Originally, [MDEV-16329 Cross-engine ALTER ONLINE TABLE](https://jira.mariadb.org/browse/MDEV-16329) proposed that InnoDB should perform character set conversions. Please read the change history of that ticket. I don’t think that we should attempt to extend the capability of InnoDB to perform conversions while rebuilding a table.

I think that we only want the following in InnoDB:
* If the column is not indexed, allow instantaneous changes of collation.
* If the collation is compatible (say, converting utf8mb3_general_ci to utf8mb4_general_ci), allow `ALGORITHM=INSTANT` always for `VARCHAR` columns. (For `CHAR` columns not for `ROW_FORMAT=REDUNDANT`, because the encoding would not be compatible if the `mbmaxlen` changes.)
* If the encoding is compatible but the collation is not, allow `ALGORITHM=NOCOPY` to drop/add the affected indexes.

There is no need to support table rebuild with `ALGORITHM=INPLACE`, for example if the `PRIMARY KEY` depends on the changed collation.